### DISCUSSION TITLE
`reportMatchNotExhaustive` false positive for tuples

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#8322] [General]
Minimal reproducible example ([playground](https://pyright-play.net/?pyrightVersion=1.1.370&pythonVersion=3.13&strict=true&code=CYUwZgBGAUA2BcEAuBXADrEBtAlgOyQBoIA6MgXQEoIBaAPgnyXgFgAoCTiAWwEMkAxgAsICdlwkQBvAM4gIWcqw6TVAJxCo1eCAAZxqqbPlYhIXsGIAqAPoaZSJQcNcNWnWYvsgA)):

```python
def f(l: tuple[int, ...]) -> int:
    match l:  # Cases within match statement do not exhaustively handle all values. Unhandled type: "tuple[int, ...]"
        case []:
            return 0
        case [head, *_rest]:
            return head
```

This is almost identical to #6843, but the reason for its closure doesn't apply here. Is this a bug?

**Comment by erictraut** [comment]
This is expected behavior, not a bug. Match exhaustion is detected via type narrowing. In your example, the first case statement (with pattern `[]`) does not narrow the subject type in the fall-through case. Its type remains `tuple[int, ...]`. For comparison, mypy's behavior is the same.

**Reply by qexat** [reply]
In the case where the first pattern is `()`, couldn't it narrow to `tuple[int, *tuple[int, ...]]`?

**Reply by qexat** [reply]
> In the case where the first pattern is `()`, couldn't it narrow to `tuple[int, *tuple[int, ...]]`?

After some thinking, I'm not sure that it would play well with more complex types, and there are probably some non-obvious edge cases that would prevent this narrowing that I haven't thought about
